api/src/main/java/net/md_5/bungee/api/event/AsyncEvent.java

```diff
@@ -43,30 +42,44 @@ public void postCall()
     /**
      * Register an intent that this plugin will continue to perform work on a
      * background task, and wishes to let the event proceed once the registered
-     * background task has completed.
+     * background task has completed. Multiple intents can be registered by a
+     * plugin, but the plugin must complete the same amount of intents for the
+     * event to proceed.
      *
      * @param plugin the plugin registering this intent
      */
     public void registerIntent(Plugin plugin)
     {
         Preconditions.checkState( !fired.get(), "Event %s has already been fired", this );
-        Preconditions.checkState( !intents.contains( plugin ), "Plugin %s already registered intent for event %s", plugin, this );
 
-        intents.add( plugin );
+        AtomicInteger intentCount = intents.get( plugin );
+
+        if ( intentCount == null )
+        {
+            intents.put( plugin, new AtomicInteger( 1 ) );
+        } else
+        {
+            intentCount.incrementAndGet();
+        }
+
         latch.incrementAndGet();
     }
 
     /**
-     * Notifies this event that this plugin has done all its required processing
-     * and wishes to let the event proceed.
+     * Notifies this event that this plugin has completed an intent and wishes
+     * to let the event proceed once all intents have been completed.
      *
      * @param plugin a plugin which has an intent registered for this event
      */
     @SuppressWarnings("unchecked")
     public void completeIntent(Plugin plugin)
     {
-        Preconditions.checkState( intents.contains( plugin ), "Plugin %s has not registered intent for event %s", plugin, this );
-        intents.remove( plugin );
+        AtomicInteger intentCount = intents.get( plugin );
+
+        Preconditions.checkState( intentCount != null && intentCount.get() > 0, "Plugin %s has not registered intents for event %s", plugin, this );
+
+        intentCount.decrementAndGet();
+
         if ( fired.get() )
         {
             if ( latch.decrementAndGet() == 0 )
```

Comment:
I think we may need to change this to `<= 0` otherwise there could be an undetectable race from a bad plugin that would hang the callback:

Thread 1: complete -> checkstate -> cnt == 1
Thread 2: complete -> checkstate -> cnt == 1
Thread 1: decrement -> cnt == 0
Thread 2: decrement -> cnt == -1

Event --> Hung


Yes the plugin is wrong, but it will never error... The perils of avoiding synchronization

Reply:
Yeah the problem is that intentCount could change from checkState to when it's actually decremented. If there was a "decrementIf" method there would be no issue. This would be a problem of the latch decrementing for intents a plugin doesn't have even if it isn't the last intent. However, I don't see how that would hang the callback? The callback is still called when the latch hits 0 and if we changed it to <= 0 it would be called multiple times. Maybe instead throw an exception/log it?

Reply:
Oh right, the latch will always go past zero.